Generate Cpp namespace when import Cpp is used

bricknerb opened the pull request:
Also defined a dedicated `ImportCppDecl` `InstKind`.
Part of #4666

bricknerb committed:
Generate `Cpp` namespace when `import Cpp` is used
Also defined a dedicated `ImportCppDecl` `InstKind`.

bricknerb committed:
Add missing comments

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -340,17 +345,22 @@ auto CheckUnit::ImportCppPackages() -> void {
 
+  IdentifierId package_id = imports.front().names.package_id;
+  for (size_t i = 1; i < imports.size(); ++i) {
+    CARBON_CHECK(imports[i].names.package_id == package_id);
```
Might be nicer to use `all` here instead of a for loop? If you like:
```
IdentifierId package_id = imports.front().names.package_id;
CARBON_CHECK(
    llvm::all(imports, [&](const CppImport& import) { return import.names.package_id == package_id; }));
```

bricknerb replied:
Done.

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,81 @@ auto ImportCppFiles(
+      first_import_decl_id, SemIR::AccessKind::Public);
+  CARBON_CHECK(inserted);
+
+  CARBON_CHECK(first_import_decl_id.has_value());
```
GetLocId() checks the same thing (via `>= 0`) so I think this isn't needed here.

bricknerb replied:
Done.

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,81 @@ auto ImportCppFiles(
+  auto import_loc_id = context.insts().GetLocId(first_import_decl_id);
+
+  auto namespace_inst = SemIR::Namespace{
+      namespace_type_id, SemIR::NameScopeId::None, first_import_decl_id};
```
I believe we'd like the fields to be named in here: https://github.com/carbon-language/carbon-lang/blob/133717cd7e5b821fa832cad2ad85be95ba47ce37/docs/project/cpp_style_guide.md?plain=1#L132
(I see import.cpp didn't name them here, but does in other cases, probably an oversight/inconsistency there.)

bricknerb replied:
Done.

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,81 @@ auto ImportCppFiles(
+      namespace_id, name_id, SemIR::NameScopeId::Package);
+  context.ReplaceInstBeforeConstantUse(namespace_id, namespace_inst);
+
+  // Note we have to get the parent scope freshly, creating the imported
```
Maybe we should just avoid the local variable entirely? It's only used once each time. And leave this comment around explaining why we don't hold a local variable instead?

bricknerb replied:
Done.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/cpp_diagnostics.carbon:
```diff
@@ -219,47 +219,75 @@ import Cpp library "multiple_errors_and_multiple_warnings.h";
+// CHECK:STDOUT:   package: <namespace> = namespace [template] {
+// CHECK:STDOUT:     .Cpp = imports.%Cpp
+// CHECK:STDOUT:   }
+// CHECK:STDOUT:   import_cpp
```
This looks different from the regular import semir. For example, stripped down from `toolchain/check/testdata/alias/fail_builtins.carbon`:

```
// CHECK:STDOUT: imports {
// CHECK:STDOUT:   %Core: <namespace> = namespace file.%Core.import, [template] {
// CHECK:STDOUT:     .Int = %Core.Int
// CHECK:STDOUT:   }
// CHECK:STDOUT: }
// CHECK:STDOUT:
// CHECK:STDOUT: file {
// CHECK:STDOUT:   package: <namespace> = namespace [template] {
// CHECK:STDOUT:     .Core = imports.%Core
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %Core.import = import Core
// CHECK:STDOUT: }
```

a) The `file` name is `%Core.import` but here we have `import_cpp`. When there's more than one `import_cpp` they all have the same name, and there's no `%` attached to it.
b) `%Core.import` is `= import Core` but there's nothing attached to `import_cpp`
c) In the `imports` block, the `%Core` is set to `namespace file.%Core.import` which matches the name we see in the `file` block. But for cpp we have `%Cpp` set to `namespace file.%cpp_import` which is not the same as `import_cpp` in `file`.

How do we make this more consistent?

bricknerb replied:
That's a good question.
I think the main difference is that when we import Cpp files, we generate a header that includes all the imported Cpp headers and process the generated header, so all Cpp imports are actually merged.
a) `import` is replaced by `import_cpp` as expected. There is no specific package name because all `import_cpp` are for `Cpp`.
b) That's because of the generated file that merges all `import Cpp`s.
c) I've replaced `cpp_import` with `import_cpp`, so it matches better.

That said, I'm looking for suggestions what should the output look like given that we merge all `import Cpp` together. Perhaps some improvements can be done to this PR and more complicated ones can be done in follow ups.

jonmeow replied:
`ImportDecl` and `ImportCppDecl` have separate printing behavior. When you see `%Core.import = import Core`, it's declaring an `ImportDecl` with `Core` as an argument, and assigning a name as `%Core.import`. Right now `CppImport` doesn't have any arguments, so nothing on the RHS should be expected.

@bricknerb In `formatter.cpp`, if you copy `auto FormatInstLHS(InstId inst_id, ImportDecl /*inst*/) -> void` to add a similar overload for `ImportCppDecl`, what's the result? I think you might want to aim for something like `%Cpp = import_cpp`, which is how we indicate the instruction has a name.

bricknerb replied:
Thanks Jon!
Added a simple overload and we now have
`%import_cpp = import_cpp`

bricknerb committed:
Use `llvm::all_of()` to `CARBON_CHECK` all cpp imports have the same package.

bricknerb committed:
Remove unecessary `CARBON_CHECK` since `GetLocId()` in the next line has an equivalent check

bricknerb committed:
Use designated initializers for `SemIR::Namespace`

bricknerb committed:
Remove `parent_scope` variable as it's only used once each time

bricknerb committed:
Replace `cpp_import` with `import_cpp`

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -118,10 +121,11 @@ static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
 
   // Note we have to get the parent scope freshly, creating the imported
   // namespace may invalidate the pointer above.
```
```suggestion
  // namespace may invalidate any pointers into name_scopes().
```

bricknerb replied:
Done.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/multiple_imports.carbon:
```diff
@@ -26,7 +26,7 @@ import Cpp library "file2.h";
 // CHECK:STDOUT:
 // CHECK:STDOUT: imports {
-// CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%cpp_import.loc4, [template] {}
+// CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%import_cpp.loc4, [template] {}
```
The location is interesting, considering they are all being merged as you said. I'm not sure what that will do for diagnostics, if it will be hard to know which import is bringing in a clang error? Or maybe that is clear in diagnostics but we just don't have them all listed in semir? Then that might make tooling difficult though.

I think this is okay for now, but maybe we can think about improvements if this is a pain. Perhaps we could have multiple %Cpp imports and they'd get `.xyz` suffixes in semir.

jonmeow replied:
This is how all imported namespaces behave -- C++ isn't any different here. i.e., you should expect the same result from:

```
import Foo library "bar";
import Foo library "baz";
```

The name `Foo` can have only one location associated with it.

When it comes to _specific_ names, we can have more specific locations. We probably won't have that short-term for C++ due to complexities involves, and how `#include` semantics need to take priority. But a single name -- whether `Cpp`, `Foo`, or something else -- has a single location in SemIR.

jonmeow commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -831,6 +831,13 @@ struct ImportDecl {
   NameId package_id;
 };
 
+// An `import Cpp` declaration.
+struct ImportCppDecl {
+  static constexpr auto Kind =
+      InstKind::ImportCppDecl.Define<Parse::ImportDeclId>(
+          {.ir_name = "import_cpp", .is_lowered = false});
+};
+
```
Minor, I see you sorted elsewhere, but `ImportCppDecl` before `ImportDecl` here too

bricknerb replied:
Done.

jonmeow reviewed (commented):
I'm trying to understand the instruction model choice here, maybe you can fill out some detail for me?

At present, it looks like each `import Cpp` line gets its own instruction. Then, these are combined into a single namespace. The import appears to only be associated with the first `import Cpp`, is that correct?

Had you considered instead having the `Cpp` namespace backed by a single instruction, and having that linked to a structure that stores the information about the `Cpp` data as part of `SemIR::File`?

Some of the significance for me is in how name references work. I'd expect SemIR for something like `Cpp.printf` to read a bit like:

```
%import_cpp = import_cpp
%printf.ref: type = name_ref printf, %import_cpp.printf [template = constants.%printf]
```

Note here that only one `ImportCppDecl` can actually end up in name lookup, and this is akin to how we use `ImportDecl` when there are multiple imports from the same package: the instruction's argument is a package, and internally it decides how to split name references across libraries.

jonmeow commented:
Note, to try to build some contrast for how imports from a package work...

```
// --- foo.carbon

package Foo library "bar";

// --- baz.carbon

package Foo library "baz";

class C {}

// --- wiz.carbon

import Foo library "bar";
import Foo library "baz";

var x: Foo.C;
```

We get the detail of the imported libraries inside the import data:

```
// CHECK:STDOUT: imports {
// CHECK:STDOUT:   %Foo: <namespace> = namespace file.%Foo.import, [template] {
// CHECK:STDOUT:     .C = %Foo.C
// CHECK:STDOUT:     import Foo//bar
// CHECK:STDOUT:     import Foo//baz
// CHECK:STDOUT:   }
```

At the file level, they're combined:

```
// CHECK:STDOUT: file {
// CHECK:STDOUT:   package: <namespace> = namespace [template] {
// CHECK:STDOUT:     .Foo = imports.%Foo
// CHECK:STDOUT:     .x = %x
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %Foo.import = import Foo
```

I think here the `imports` is gotten by adding to `import_ref_ids()` -- see import.cpp's `AddNamespace`, you might want to see if you can take advantage of that rather than writing a copy.

jonmeow commented:
> Had you considered instead having the `Cpp` namespace backed by a single instruction, and having that linked to a structure that stores the information about the `Cpp` data as part of `SemIR::File`?

BTW, to correct this and connect to the cross-package example... I think a good goal would be something like:

```
// CHECK:STDOUT: imports {
// CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%Cpp.import_cpp, [template] {
// CHECK:STDOUT:     import_cpp "bar.h"
// CHECK:STDOUT:     import_cpp "baz.h"
// CHECK:STDOUT:   }
```

```
// CHECK:STDOUT: file {
// CHECK:STDOUT:   package: <namespace> = namespace [template] {
// CHECK:STDOUT:     .Cpp = imports.%Cpp
// CHECK:STDOUT:     .x = %x
// CHECK:STDOUT:   }
// CHECK:STDOUT:   %Cpp.import_cpp = import_cpp
```

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,85 @@ auto ImportCppFiles(
+  return ast;
+}
+
+static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
+                         SemIR::InstId first_import_decl_id) -> void {
```
Had you tried something like using `import.cpp`'s `AddNamespace`, i.e.:

```
AddNamespace(context, namespace_type_id, name_id, SemIR::NameScopeId::Package, /*diagnose_duplicate_namespace=*/true, [&] { return first_import_decl_id; });
```

What was the difference between that and what you're trying to do here?

bricknerb replied:
I've changed it further, but I still believe you're right that there are a lot of similarities and that's why I have a TODO extract and deduplicate the common logic.
Once we agree on the functionality, I can do this TODO as a follow up PR or as part of this PR.

jonmeow replied:
I think it's probably worth handling here... what's the effect of replacing lines 90-128 with the call?

bricknerb replied:
Sorry for missing this.
I've taken `AddNamespace()`, exposed it in `import.h` as `AddImportNamespace()` and used it in `import_cpp.cpp`.
I can put it elsewhere so I didn't see a great candidate except creating a dedicated file. WDYT?

It seems to work fine though there are some impact on the behavior in api and impl use case, due to how duplicated namespace are handled (which might be a bug fix?).

See https://github.com/carbon-language/carbon-lang/pull/4873/commits/5c8888b48939a6b8b298ebe969412d2305c92790

jonmeow replied:
I think `import.*` is still the right home for the function; the import_cpp -> import dependency is good.

Yes, I think this is pushing in the right direction for merging imports. Note, you might want a TODO to handle merging `api` and `impl` ASTs... there are a few options there though.

bricknerb replied:
Why is this dependency good? Seems like the new utility function can be extracted and used by the implementations of the two APIs, so we won't need to have one API depend on the other.

I think merging AST (and a TODO) belongs to a PR that actually tries to save the ASTs for later usage.

bricknerb committed:
Put `ImportCppDecl` before `ImportDecl`

bricknerb committed:
Improve invalidation comment

bricknerb committed:
Add `FormatInstLHS()` overload for `ImportCppDecl`

jonmeow commented on toolchain/sem_ir/inst_namer.cpp:
```diff
@@ -705,6 +705,10 @@ auto InstNamer::CollectNamesInBlock(ScopeId top_scope_id,
         continue;
       }
+      case ImportCppDecl::Kind: {
+        add_inst_name("import_cpp");
```
```suggestion
        add_inst_name("Cpp.import_cpp");
```

We're typically trying to reflect the identifier as part of instruction names so that the parallels with code are easier to see. How about this, which is close to the `ImportDecl` format?

(noting this from the `%import_cpp = import_cpp` reply, this should make it `%Cpp.import_cpp = import_cpp`)

bricknerb replied:
Done.

bricknerb committed:
Instead of creating `ImportCppDecl` per `import Cpp`, create a single one for the `Cpp` namespace.
Create a new `ImportCpp` struct that contains information per `import Cpp`, and store all instances in `File`.
Format `ImportCppDecl` by printing all of the file's `ImportCpp`.
Name `ImportCppDecl` as `"Cpp.import_cpp"` instead of `"import_cpp"`.

bricknerb commented:
> > Had you considered instead having the `Cpp` namespace backed by a single instruction, and having that linked to a structure that stores the information about the `Cpp` data as part of `SemIR::File`?
> 
> BTW, to correct this and connect to the cross-package example... I think a good goal would be something like:
> 
> ```
> // CHECK:STDOUT: imports {
> // CHECK:STDOUT:   %Cpp: <namespace> = namespace file.%Cpp.import_cpp, [template] {
> // CHECK:STDOUT:     import_cpp "bar.h"
> // CHECK:STDOUT:     import_cpp "baz.h"
> // CHECK:STDOUT:   }
> ```
> 
> ```
> // CHECK:STDOUT: file {
> // CHECK:STDOUT:   package: <namespace> = namespace [template] {
> // CHECK:STDOUT:     .Cpp = imports.%Cpp
> // CHECK:STDOUT:     .x = %x
> // CHECK:STDOUT:   }
> // CHECK:STDOUT:   %Cpp.import_cpp = import_cpp
> ```

Thanks Jon!

I've changed the logic to have a single `ImportCppDecl` instruction and have that is pointed from the imported namespace.
This instruction is in file and I added metadata in `File` for the list of cpp files imported and format them as part of that (so under `file {}` and not under `imports {}`.
WDYT?

bricknerb committed:
ClangTidy fix

danakj commented on toolchain/sem_ir/ids.h:
```diff
@@ -377,6 +378,16 @@ struct GenericInstIndex : public IndexBase<GenericInstIndex> {
+  using ValueType = ImportCpp;
+
+  // An ID with no value.
+  static const ImportCppId None;
```
The struct should be followed with
```
constexpr ImportCppId ImportCppId::None = ImportCppId(NoneIndex);
```

bricknerb replied:
Done.

bricknerb committed:
Initialize `ImportCppId::None`

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -1078,6 +1084,20 @@ class FormatterImpl {
+      out_ << "import Cpp \""
+           << FormatEscaped(
+                  sem_ir_->string_literal_values().Get(import_cpp.library_id))
+           << "\" = " << SemIR::LocId(import_cpp.node_id) << "\n";
```
```suggestion
           << "\"\n";
```

As mentioned this morning, we're only putting locations in semir when it's required to disambiguate names. i.e., if there are two instructions get named `%Foo`, we might disambiguate the second as `%Foo.loc4_1` (specifically, I think the first still remains `%Foo` and later duplicates are get unique-ified)

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.h:
```diff
@@ -13,10 +13,10 @@ namespace Carbon::Check {
-    Context& context, llvm::StringRef importing_file_path,
-    llvm::ArrayRef<std::pair<llvm::StringRef, SemIRLoc>> imports,
-    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs) -> void;
+// If successful, adds a 'Cpp' namespace.
```
Wrapping, and prefer backticks instead of quotes (GitHub doesn't support a suggestion here):

```
// report errors and warnings.
// If successful, adds a 'Cpp' namespace.
```

->

```
// report errors and warnings. If successful, adds a `Cpp` namespace.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -22,28 +23,26 @@ namespace Carbon::Check {
+static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
+                        llvm::ArrayRef<Parse::Tree::PackagingNames> imports,
+                        llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs)
+    -> std::optional<std::unique_ptr<clang::ASTUnit>> {
```
Given this is a pointer, had you considered using nullptr instead of  `std::optional`? I think that's what we're doing elsewhere, because it reduces the amount of type wrapping. We generally try not to assume that pointers are non-null, and only use `optional` when there's not another way to indicate "none".

```suggestion
    -> std::unique_ptr<clang::ASTUnit> {
```

But more than that, is it worth trying to always return an AST? i.e., even if the C++ code has an error, do you think it's worth trying to emit diagnostics where Carbon code touches C++ code which successfully produced AST nodes, as part of error recovery during diagnosis?

bricknerb replied:
I've changed that to always return an AST if Clang managed to generate one, even if there are errors.
If we find this is cause more harm than good we can change that approach.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -22,28 +23,26 @@ namespace Carbon::Check {
-    return;
-  }
-
+static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
```
Style recommends commenting functions. Maybe:

```suggestion
// Returns an AST for the C++ imports, or null on error.
static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,85 @@ auto ImportCppFiles(
+    IdentifierId cpp_package_id, SemIR::InstId first_import_decl_id,
+    llvm::ArrayRef<std::pair<llvm::StringRef, SemIRLoc>> imports,
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs) -> void {
+  size_t num_imports = imports.size();
+  if (num_imports == 0) {
```
```suggestion
  if (imports.empty()) {
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,85 @@ auto ImportCppFiles(
+  }
+
+  auto ast = GenerateAst(context, importing_file_path, imports, fs);
+  if (!ast) {
+    return;
+  }
```
We should probably do some error recovery here so that Carbon code can be checked as best as possible, even when there are C++ errors. Note this may mean we should try to produce an AST, even when that AST has errors.

Even if an AST cannot be produced, we can still do some error recovery with `scope.set_has_error()`. That should cause name lookups to silently fail, versus the errors that result from declining to add a namespace.

As an example of the behavior goal:

```
import Cpp library "error.h"

// Calling `scope.set_has_error()` should make this silently fail.
// At present, I would expect a `Cpp` name lookup error.
var x: Cpp.Foo;

// This should still diagnose the type mismatch, because it doesn't touch `Cpp`.
var x: i32 = 1.0;
```

So for example here:

```
auto ast = GenerateAst(...);
auto name_scope_id = AddNamespace(...);
if (!ast) {
  context.name_scopes().Get(name_scope_id).set_has_error();
}
```

bricknerb replied:
ok, I now always returns the pointer to the AST and mark the scope with error if it's nullptr.
We can still have errors when generating the AST, but if it's created it would not be marked with errors.
I'm not sure whether we want to mark the scope with `has_error` if Clang triggered errors but some AST was created, so I took the conservative (and simpler) approach.

jonmeow commented on toolchain/check/testdata/interop/cpp/no_prelude/cpp_namespace.carbon:
```diff
@@ -0,0 +1,137 @@
+// CHECK:STDERR: ^~~~~~
+// CHECK:STDERR:
+class Cpp.C {};
+
```
Can you add another test that does `import Cpp` in an `impl` file:

```suggestion

// --- api_and_impl.carbon

library "[[@TEST_NAME]]";

import Cpp library "header.h";

// --- api_and_impl.impl.carbon

impl library "[[@TEST_NAME]]";

import Cpp library "header.h";
```

bricknerb replied:
Nice catch!
Indeed in this case we try to add the `Cpp` name to `Package` scope twice.
I think I've fixed that.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,13 +68,85 @@ auto ImportCppFiles(
+  return ast;
+}
+
+static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
```
```suggestion
// Adds a namespace for the `Cpp` import.
static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
```

bricknerb replied:
Done.

jonmeow reviewed (commented):
The IR looks good to me, and I also checked with zygoloid to ensure it makes sense to him. The one comment I have there is to drop the location, which I think I'd also talked when we met.

Some other code comments though...

bricknerb committed:
Remove location from `import Cpp` formatting

bricknerb committed:
Add and reformat comments.
Use `.empty()`.

bricknerb committed:
When generating the AST triggers Clang errors, continue to create the `Cpp` namespace and mark its `NameScope` with error

bricknerb committed:
Support `import Cpp` in both API and Impl

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -77,6 +79,80 @@ auto ImportCppFiles(
+        return import.package_id == package_id;
+      }));
+  auto name_scope_id = AddNamespace(context, package_id, imports);
+
+  if (!ast) {
+    context.name_scopes().Get(name_scope_id).set_has_error();
+  }
```
I think the separate changes to `GenerateAST` mean an `ast` will always be non-null here. Note though, you could make `GenerateAst` return something like `std::pair<Ast, bool has_error>` (i.e., to communicate back the "error" state while still returning an AST). So I think you probably want one of:

```suggestion
```

```suggestion

  if (ast_has_error) {
    context.name_scopes().Get(name_scope_id).set_has_error();
  }
```

bricknerb replied:
Done.
Note that `ast` could be null, but not due to errors.

jonmeow replied:
Sorry, I'm definitely missing this. Can you document the return value of `GenerateAst`, both to explain the `bool` and to explain when the AST will be null?

jonmeow replied:
Digging further, it looks like you're talking about buildASTFromCodeWithArgs can return null if Invocation.run returns false. I wonder if this makes the `has_errors` check redundant right now... but also, maybe it means that we'll want to make our own version that returns an AST even when there are errors?

To be sure, not asking for changes to that here, just trying to make sure my understanding is correct -- maybe something for a TODO.

bricknerb replied:
Elaborated on the return value and added a TODO.
I believe that we should re-evaluate once we have a serialized AST.

jonmeow reviewed (commented):
Basically LG, but I want to highlight https://github.com/carbon-language/carbon-lang/pull/4873#discussion_r1943687041 since you pinged for another review and might've missed it.

bricknerb committed:
Change `GenerateAst()` to return whether errors were encountered and use that to set `has_error` in `NameScope`

bricknerb committed:
Reuse `import.cpp`'s `AddNamespace()` in `import_cpp.cpp`'s `AddNamespace()`.

bricknerb committed:
Add a parameter comment

bricknerb commented:
> Basically LG, but I want to highlight [#4873 (comment)](https://github.com/carbon-language/carbon-lang/pull/4873#discussion_r1943687041) since you pinged for another review and might've missed it.

Thanks!
Sorry for missing this one.

jonmeow commented on toolchain/check/import.cpp:
```diff
@@ -617,4 +541,68 @@ auto ImportNameFromOtherPackage(
   return result_id;
 }
 
+auto AddImportNamespace(Context& context, SemIR::TypeId namespace_type_id,
```
I can't view a delta because you moved the code. Can you please clarify why this needed to be moved, or to help me in review (and maintain blame history), move it back?

bricknerb replied:
I've moved it to match the order of the methods in the .h.
I'm happy to try and do this move in a separate PR or move the entire refactoring to another PR if you think it's better.
It might make sense to move it to a completely separate .h/.cpp.
Eventually the code is likely to be moved unless we're preferring blame history over code style.
Let me know what you prefer.

jonmeow replied:
> I've moved it to match the order of the methods in the .h.

How about doing this the other way, putting it in the .h in a spot that matches the order in the cpp file? i.e., don't put it last, put it first. Then I think we can get the best of both worlds: order in both would match, and it wouldn't need to move - in this PR or any other.

bricknerb replied:
Done.
However, I feel this should not be the final location.
1. IMO, iIt makes more sense to put this function declared down because it's not the primary `import.h` API (and perhaps should actually move to a different file). BTW, when moving code to a different file, can we make it easy to review or keep blame history? Do we usually do that? Do we avoid moving code between files because of that?
2. The definition is now between internal linkage functions instead of being close to the API functions. I don't think this is explicitly against the style guide, but definitely seems inconsistent.

Should I move the code to a better location (after you review or in a separate PR)?

jonmeow replied:
We can discuss this separately, but as I've said, I think import.h is the right home because it's core import functionality.

jonmeow commented on toolchain/check/import.h:
```diff
@@ -11,6 +11,12 @@
 
 namespace Carbon::Check {
 
+struct AddImportNamespaceResult {
+  SemIR::NameScopeId name_scope_id;
+  SemIR::InstId inst_id;
+  bool is_duplicate_of_namespace_in_current_package;
+};
```
Thoughts on moving this to be adjacent to `AddImportNamespace`? Since it's a free function (vs a class method where there's declaration order to consider), I think it'd be helpful to have the return struct and function be adjacent.

bricknerb replied:
Done.

bricknerb committed:
Move `AddImportNamespaceResult` closer to where it's used.

bricknerb committed:
Move `AddImportNamespace()` to the beginning of `import.h` and back to its original location in `import.cpp` for easier reviewing

bricknerb committed:
Elaborate on the return value of `GenerateAst()`

bricknerb committed:
Add a TODO to always return an AST

bricknerb committed:
Merge remote-tracking branch 'upstream/trunk' into interop

bricknerb committed:
Merge remote-tracking branch 'upstream/trunk' into interop

bricknerb committed:
In `import_cpp`, use `PackageNameId` instead of `IdentifierId` for package ids following #4909.
